extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/OpenTelemetryAgroalDataSource.java

```diff
@@ -0,0 +1,66 @@
+package io.quarkus.agroal.runtime;
+
+import java.sql.SQLException;
+import java.sql.ShardingKeyBuilder;
+import java.util.Collection;
+import java.util.List;
+
+import io.agroal.api.AgroalDataSource;
+import io.agroal.api.AgroalDataSourceMetrics;
+import io.agroal.api.AgroalPoolInterceptor;
+import io.agroal.api.configuration.AgroalDataSourceConfiguration;
+import io.opentelemetry.api.GlobalOpenTelemetry;
+import io.opentelemetry.instrumentation.jdbc.datasource.OpenTelemetryDataSource;
+
+/**
+ * The {@link AgroalDataSource} wrapper that actives OpenTelemetry JDBC instrumentation.
+ */
+public class OpenTelemetryAgroalDataSource extends OpenTelemetryDataSource implements AgroalDataSource {
+
+    private final AgroalDataSource delegate;
+
+    public OpenTelemetryAgroalDataSource(AgroalDataSource delegate) {
+        super(delegate, GlobalOpenTelemetry.get());
```

Comment:
We must not use GlobalOpenTelemetry.get().
I wonder if we don't need to send this `AgroalDataSource` to the OTel extension and wire `OpenTelemetryAgroalDataSource` there.
We need to have a way to set the OpenTelemetry object later.

Reply:
I realize that, but I thought it would be the easiest to refactor with your changes in #30033. Anyway, point taken, I'll read up #30033 and prepare solution that will not require changes later.

Reply:
I'll introduce OT SPI and move things around.

Reply:
What you suggested works for me. No worries.

Reply:
Thanks @brunobat , right now we can't really force order that first, OpenTelemetry is created and then step that set up datasources and uses OTEL runs because (Tracer ->) OpenTelemetry must run after CDI bean container is fully initialized, while Agroal step produces additional bean [circular reference].

I think best solution would be to refactor it based on your PR (now everything works as OTEL is initialized during static init). I'm happy to do that, just not here as you would hit merge hell.